lightning/src/ln/chan_utils.rs

```diff
@@ -690,25 +781,305 @@ impl HolderCommitmentTransaction {
+		let sighash = &bip143::SigHashCache::new(&self.transaction).signature_hash(0, funding_redeemscript, channel_value_satoshis, SigHashType::All)[..];
+		hash_to_message!(sighash)
+	}
+
+	/// Sign a transaction, either because we are counter-signing the counterparty's transaction or
+	/// because we are about to broadcast a holder transaction.
+	pub fn sign<T: secp256k1::Signing + secp256k1::Verification>(&self, funding_key: &SecretKey, funding_redeemscript: &Script, channel_value_satoshis: u64, secp_ctx: &Secp256k1<T>) -> Signature {
+		let sighash = self.get_sighash_all(funding_redeemscript, channel_value_satoshis);
+		secp_ctx.sign(&sighash, funding_key)
+	}
+}
+
+/// This class tracks the per-transaction information needed to build a commitment transaction and to
+/// actually build it and sign.  It is used for holder transactions that we sign only when needed
+/// and for transactions we sign for the counterparty.
+///
+/// This class can be used inside a signer implementation to generate a signature given the relevant
+/// secret key.
+#[derive(Clone)]
+pub struct CommitmentTransaction {
+	commitment_number: u64,
+	to_broadcaster_value_sat: u64,
+	to_countersignatory_value_sat: u64,
+	feerate_per_kw: u32,
+	htlcs: Vec<HTLCOutputInCommitment>,
+	// A cache of the parties' pubkeys required to construct the transaction
+	keys: TxCreationKeys,
+	// For access to the pre-built transaction, see doc for untrusted_built_transaction
+	built: BuiltCommitmentTransaction,
+}
+
+impl PartialEq for CommitmentTransaction {
+	fn eq(&self, o: &Self) -> bool {
+		return self.commitment_number == o.commitment_number &&
+			self.to_broadcaster_value_sat == o.to_broadcaster_value_sat &&
+			self.to_countersignatory_value_sat == o.to_countersignatory_value_sat &&
+			self.feerate_per_kw == o.feerate_per_kw &&
+			self.htlcs == o.htlcs &&
+			self.keys == o.keys &&
+			self.check_built_transaction(o);
+	}
+}
+
+impl Writeable for Vec<HTLCOutputInCommitment> {
+	#[inline]
+	fn write<W: Writer>(&self, w: &mut W) -> Result<(), ::std::io::Error> {
+		(self.len() as u16).write(w)?;
+		for e in self.iter() {
+			e.write(w)?;
+		}
+		Ok(())
+	}
+}
+
+impl Readable for Vec<HTLCOutputInCommitment> {
+	#[inline]
+	fn read<R: Read>(r: &mut R) -> Result<Self, DecodeError> {
+		let len: u16 = Readable::read(r)?;
+		let byte_size = (len as usize)
+			.checked_mul(HTLC_OUTPUT_IN_COMMITMENT_SIZE)
+			.ok_or(DecodeError::BadLengthDescriptor)?;
+		if byte_size > MAX_BUF_SIZE {
+			return Err(DecodeError::BadLengthDescriptor);
+		}
+		let mut ret = Vec::with_capacity(len as usize);
+		for _ in 0..len { ret.push(HTLCOutputInCommitment::read(r)?); }
+		Ok(ret)
+	}
+}
+
+impl_writeable!(CommitmentTransaction, 0, {
+	commitment_number,
+	to_broadcaster_value_sat,
+	to_countersignatory_value_sat,
+	feerate_per_kw,
+	htlcs,
+	keys,
+	built
+});
+
+impl CommitmentTransaction {
+	/// Construct an object of the class while assigning transaction output indices to HTLCs.
```

Comment:
Something something "only include HTLCs which are above the dust limit for this transaction".

Reply:
done